lens-cube/src/main/java/org/apache/lens/cube/parse/CandidateCoveringSetsResolver.java

```diff
@@ -82,73 +82,80 @@ private void updateFinalCandidates(List<List<UnionCandidate>> jcs) {
+      }
+    }
+  }
+
+  private void resolveRangeCoveringFactSet(CubeQueryContext cubeql, List<TimeRange> ranges,
                                            Set<QueriedPhraseContext> queriedMsrs) throws LensException {
-    // All facts
-    List<Candidate> allCandidates = cubeql.getCandidateSet();
-    // Partially valid facts
+    // All Candidates
+    List<Candidate> allCandidates = new ArrayList<Candidate>(cubeql.getCandidates());
+    // Partially valid candidates
     List<Candidate> allCandidatesPartiallyValid = new ArrayList<>();
     for (Candidate cand : allCandidates) {
       // Assuming initial list of candidates populated are StorageCandidate
       if (cand instanceof StorageCandidate) {
         StorageCandidate sc = (StorageCandidate) cand;
-        if (CandidateUtil.isValidForTimeRange(sc, range)) {
-          List<Candidate> one = new ArrayList<Candidate>(Arrays.asList(sc));
-          unionCandidates.add(new UnionCandidate(one));
-        } else if (CandidateUtil.isPartiallyValidForTimeRange(sc, range)) {
-          allCandidatesPartiallyValid.add(sc);
+        if (CandidateUtil.isValidForTimeRanges(sc, ranges)) {
+          List<Candidate> one = new ArrayList<Candidate>(Arrays.asList(CandidateUtil.cloneStorageCandidate(sc)));
+          unionCandidates.add(new UnionCandidate(one, "uc" + unionCandidatealiasCounter++));
+          continue;
+        } else if (CandidateUtil.isPartiallyValidForTimeRanges(sc, ranges)) {
+          allCandidatesPartiallyValid.add(CandidateUtil.cloneStorageCandidate(sc));
         }
       } else {
         throw new LensException("Not a StorageCandidate!!");
       }
     }
     // Get all covering fact sets
-    List<List<Candidate>> coveringFactSets =
+    List<UnionCandidate> unionCoveringSet =
         getCombinations(new ArrayList<Candidate>(allCandidatesPartiallyValid));
-    // Sort the Collection based on size
-    Collections.sort(coveringFactSets, new Comparator<List<?>>() {
-      @Override
-      public int compare(List<?> o1, List<?> o2) {
-        return Integer.valueOf(o1.size()).compareTo(o2.size());
-      }
-    });
-
-    // iterate over the candidate set and remove the one which can't answer the range
-    for (Iterator<List<Candidate>> itr = coveringFactSets.iterator(); itr.hasNext(); ) {
-      List<Candidate> cand = itr.next();
-      if (!CandidateUtil.isTimeRangeCovered(cand, range.getFromDate(), range.getToDate())) {
-        itr.remove();
-      }
-    }
+    // Sort the Collection based on no of elements
+    Collections.sort(unionCoveringSet, new UnionCandidateComparator<UnionCandidate>());
+    // prune non covering sets
+    pruneUnionCandidatesNotCoveringAllRanges(unionCoveringSet, ranges);
     // prune candidate set without common measure
-    pruneCoveringSetWithoutAnyCommonMeasure(coveringFactSets, queriedMsrs, cubeql);
+    pruneUnionCoveringSetWithoutAnyCommonMeasure(unionCoveringSet, queriedMsrs, cubeql);
     // prune redundant covering sets
```

Comment:
Can we add more details in comment here to clarity what we intend to do in pruneRedundantUnionCoveringSets